### PR TITLE
TLS group handling revision.

### PR DESCRIPTION
**Pull request by snhenson**
Further revision of the group id handling code. This tidies up a number of areas in the group handling mainly in t1_lib.c and removes some code duplication.

Extended tests hopefully added this time ;-)

### CODE REVIEW

File: ssl/s3_lib.c [hunk 1]
```diff
@@ -3633,8 +3634,16 @@ long ssl3_ctrl(SSL *s, int cmd, long larg, void *parg)
                                     &s->ext.supportedgroups_len, parg);
 
     case SSL_CTRL_GET_SHARED_GROUP:
-        return tls1_shared_group(s, larg);
+        {
+            uint16_t id = tls1_shared_group(s, larg);
+            if (larg != -1) {
```
Comment (richsalz) [inline ssl/s3_lib.c]: blank line

File: ssl/s3_lib.c [hunk 2]
```diff
@@ -3633,8 +3634,16 @@ long ssl3_ctrl(SSL *s, int cmd, long larg, void *parg)
...
+                const TLS_GROUP_INFO *ginf = tls1_group_id_lookup(id);
 
+                return ginf == NULL ? 0 : ginf->nid;
+            } else {
```
Comment (richsalz) [inline ssl/s3_lib.c]: fwiw i hate else after return.  not a blocking comment. :)

File: ssl/s3_lib.c [hunk 3]
```diff
@@ -4612,6 +4621,41 @@ EVP_PKEY *ssl_generate_pkey_curve(int id)
     EVP_PKEY_CTX_free(pctx);
     return pkey;
 }
+
+/* Generate a param from a group ID */
```
Comment (richsalz) [inline ssl/s3_lib.c]: I think the multi-line format for comments before a function makes it pop out more.  not a blocking comment :)

File: ssl/ssl_locl.h [hunk 4]
```diff
@@ -1349,6 +1349,19 @@ typedef struct sigalg_lookup_st {
     int curve;
 } SIGALG_LOOKUP;
 
+typedef struct {
```
Comment (richsalz) [inline ssl/ssl_locl.h]: a name for the struct, like tls_group_info_st ?

File: ssl/ssl_locl.h [hunk 5]
```diff
@@ -1349,6 +1349,19 @@ typedef struct sigalg_lookup_st {
...
+    int nid;                    /* Curve NID */
+    int secbits;                /* Bits of security (from SP800-57) */
+    uint16_t flags;             /* Flags: currently just group type */
+} TLS_GROUP_INFO;
+
+/* flags values */
+/* Mask for group type */
```
Comment (richsalz) [inline ssl/ssl_locl.h]: this second comment confused me at first.  maybe indent or move it onto the define line?

**Comment by snhenson** [comment]
Addressed comments, please reconfirm +1.

**Comment by richsalz** [comment]
reconfirming.  thanks

**Comment by snhenson** [comment]
Thanks, pushed.

**Pull request closed without merging**